src/renderer/gdi/paint.cpp

```diff
@@ -543,20 +543,24 @@ bool GdiEngine::FontHasWesternScript(HDC hdc)
     };
     const auto DrawCurlyLine = [&](const auto x, const auto y, const auto cCurlyLines) {
         const auto curlyLineWidth = fontWidth;
-        const auto curlyLineHalfWidth = lrintf(curlyLineWidth / 2.0f);
-        const auto controlPointHeight = gsl::narrow_cast<long>(std::floor(3.5f * _lineMetrics.curlylinePeakHeight));
+        const auto curlyLineHalfWidth = std::lround(curlyLineWidth / 2.0f);
+        const auto controlPointHeight = std::lround(3.5f * _lineMetrics.curlylinePeakHeight);
+
         // Each curlyLine requires 3 `POINT`s
         const auto cPoints = gsl::narrow<DWORD>(3 * cCurlyLines);
         std::vector<POINT> points;
         points.reserve(cPoints);
+
+        const auto end = x + gsl::narrow_cast<til::CoordType>(cCurlyLines * curlyLineWidth);
         auto start = x;
-        for (auto i = 0u; i < cCurlyLines; i++)
+        while (start < end)
         {
             points.emplace_back(start + curlyLineHalfWidth, y - controlPointHeight);
             points.emplace_back(start + curlyLineHalfWidth, y + controlPointHeight);
             points.emplace_back(start + curlyLineWidth, y);
             start += curlyLineWidth;
         }
+
         RETURN_HR_IF(E_FAIL, !MoveToEx(_hdcMemoryContext, x, y, nullptr));
         RETURN_HR_IF(E_FAIL, !PolyBezierTo(_hdcMemoryContext, points.data(), cPoints));

```

Comment:
BTW in the future we might want to consider to set the DPI of the GDI context to 96. That way we could draw with pixel-accuracy instead of this buggy rounding we currently have to endure at fractual display scalings.

Reply:
Side question, isn't it already?

I can see GDI renderer is being initialized with `USER_DEFAULT_SCREEN_DPI` (through `s_iBaseDpi`)

Reply:
Oh yeah, I think you're right.